File: python/trackc++/elements.cpp
// TRACKC++
// ========
// Author: 		Ximenes R. Resende
// email:  		xresende@gmail.com, ximenes.resende@lnls.br
// affiliation:	LNLS - Laboratorio Nacional de Luz Sincrotron
// Date: 		Tue Dec 10 17:57:20 BRST 2013


#include "auxiliary.h"
#include "elements.h"

const std::vector<double> Element::default_polynom = std::vector<double>(3,0);


// default constructor (constructs a drift element)
Element::Element(const std::string& fam_name_, const double& length_) :
	fam_name(fam_name_), pass_method(PassMethod::pm_drift_pass),
	nr_steps(1), length(length_),
	hkick(0), vkick(0),
	angle(0), angle_in(0), angle_out(0),
	gap(0), fint_in(0), fint_out(0),
	thin_KL(0), thin_SL(0),
	frequency(0), voltage(0), energy(0),
	//err_dx(0), err_dy(0), err_excit(0),
	//err_roll(0), err_yaw(0), err_pitch(0),
	polynom_a(default_polynom), polynom_b(default_polynom)
{
	//polynom_a.clear();
	//polynom_b.clear();
	//t1.set_zero(); t2.set_zero();
	//r1.set_identity(); r2.set_identity();
	for(unsigned int i=0; i<6; i++) {
		t_in[i] = t_out[i] = 0.0;
		for(unsigned int j=0; j<6; ++j) {
			if (i == j) {
				r_in[i*6+j] = r_out[i*6+j] = 1.0;
			} else {
				r_in[i*6+j] = r_out[i*6+j] = 0.0;
			}
		}
	}

};

Element Element::drift (const std::string& fam_name_, const double& length_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_drift_pass;
	return e;
}


Element Element::quadrupole (const std::string& fam_name_, const double& length_, const double& K_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_str_mpole_symplectic4_pass;
	e.polynom_b[1] = K_;
	return e;
}

Element Element::sextupole (const std::string& fam_name_, const double& length_, const double& S_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_str_mpole_symplectic4_pass;
	e.polynom_b[2] = S_;
	return e;
}
Element Element::rbend (const std::string& fam_name_, const double& length_, const double& angle_, const double& angle_in_, const double& angle_out_, const double& K_, const double& S_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_bnd_mpole_symplectic4_pass;
	e.angle = angle_;
	e.angle_in = angle_in_;
	e.angle_out = angle_out_;
	e.polynom_b[1] = K_;
	e.polynom_b[2] = S_;
	return e;
}

Element Element::rfcavity (const std::string& fam_name_, const double& length_, const double& frequency_, const double& voltage_, const double& energy_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_cavity_pass;
	e.frequency = frequency_;
	e.voltage = voltage_;
	e.energy = energy_;
	return e;
}



void print_polynom(std::ostream& out, const std::string& label, const std::vector<double>& polynom) {
	int order = 0;
	for(unsigned int i=0; i<polynom.size(); ++i) {
		if (polynom[i] != 0) order = i+1;
	}
	if (order > 0) out << label;
	for(int i=0; i<order; ++i) {
		out << polynom[i] << " ";
	}
	if (order > 0) out << std::endl;
}

std::ostream& operator<< (std::ostream &out, Element& el) {

	out << "FamName      : " << el.fam_name << std::endl;
	if (el.length != 0) out << "Length       : " << el.length << std::endl;
	out << "PassMethod   : " << passmethods[el.pass_method] << std::endl;
	if (el.nr_steps > 1) out << "NrSteps      : " << el.nr_steps << std::endl;
	if (el.thin_KL != 0)  out << "ThinKL       : " << el.thin_KL << std::endl;
	if (el.thin_SL != 0)  out << "ThinSL       : " << el.thin_SL << std::endl;
	if (el.angle != 0)  out << "BendingAngle : " << el.angle << std::endl;
	if (el.angle != 0)  out << "EntranceAngle: " << el.angle_in << std::endl;
	if (el.angle != 0)  out << "ExitAngle    : " << el.angle_out << std::endl;
	if ((el.gap != 0) and ((el.fint_in != 0) or (el.fint_out != 0))) {
		out << "Gap          : " << el.gap << std::endl;
		out << "FInt1        : " << el.fint_in << std::endl;
		out << "FInt2        : " << el.fint_out << std::endl;
	}
	print_polynom(out, "PolynomA     : ", el.polynom_a);
	print_polynom(out, "PolynomB     : ", el.polynom_b);
	if (el.frequency != 0) out << "Frequency    : " << el.frequency << std::endl;
	if (el.voltage != 0)   out << "Voltage      : " << el.voltage << std::endl;
	if (el.energy != 0)    out << "Frequency    : " << el.energy << std::endl;
	return out;
}

File: pyring/elements.py
import pyring.passmethods as passmethods

class marker(object):
    def __init__(self,
         fam_name,
         pass_method = passmethods.pm_drift_pass,
         length      = 0.0,
         nr_steps    = 1,
         hkick       = None,
         vkick       = None,
         angle       = None,
         angle_in    = None,
         angle_out   = None,
         gap         = None,
         fint_in     = None,
         fint_out    = None,
         thin_KL     = None,
         thin_SL     = None,
         frequency   = None,
         voltage     = None,
         polynom_a   = None,
         polynom_b   = None,
         kicktable   = None,
         hmax        = None,
         vmax        = None,
         t_in        = None,
         t_out       = None,
         r_in        = None,
         r_out       = None,
         **kwargs
         ):

        if angle       is not None: self.angle       = angle
        if angle_in    is not None: self.angle_in    = angle_in
        if angle_out   is not None: self.angle_out   = angle_out
        if gap         is not None: self.gap         = gap
        if fint_in     is not None: self.fint_in     = fint_in
        if fint_out    is not None: self.fint_out    = fint_out
        if thin_KL     is not None: self.thin_KL     = thin_KL
        if thin_SL     is not None: self.thin_SL     = thin_SL
        if frequency   is not None: self.frequency   = frequency
        if voltage     is not None: self.voltage     = voltage
        if polynom_a   is not None: self.polynom_a   = polynom_a
        if polynom_b   is not None: self.polynom_b   = polynom_b
        if kicktable   is not None: self.kicktable   = kicktable
        if hmax        is not None: self.hmax        = hmax
        if vmax        is not None: self.vmax        = vmax
        if t_in        is not None: self.t_in        = t_in
        if t_out       is not None: self.t_out       = t_out
        if r_in        is not None: self.r_in        = r_in
        if r_out       is not None: self.r_out       = r_out

    def __str__(self):

        r  = ''
        r +=   '{0:<15s} {1}'.format('fam_name',    self.fam_name)
        r += '\n{0:<15s} {1}'.format('pass_method', self.pass_method)
        r += '\n{0:<15s} {1}'.format('length[m]',   str(self.length))
        r += '\n{0:<15s} {1}'.format('nr_steps',    str(self.nr_steps))
        try: r += '\n{0:<15s} {1}'.format('hkick[rad]',     str(self.hkick))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('vkick[rad]',     str(self.vkick))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('angle[rad]',     str(self.angle))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('angle_in[rad]',  str(self.angle_in))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('angle_out[rad]', str(self.angle_out))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('gap[m]',         str(self.gap))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('fint_in',        str(self.fint_in))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('fint_out',       str(self.fint_out))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('thin_KL[1/m]',   str(self.thin_KL))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('thin_SL[1/m^2]', str(self.thin_SL))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('frequency[Hz]',  str(self.frequency))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('voltage[V]',     str(self.voltage))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('polynom_a[m^n]', str(self.polynom_a))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('polynom_b[m^n]', str(self.polynom_b))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('kicktable',      str(self.kicktable))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('hmax[m]',        str(self.hmax))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('vmax[m]',        str(self.vmax))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('t_in[m,rad]',    str(self.t_in))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('t_out[m,rad]',   str(self.t_out))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('r_in[1,m,1/m]',  str(self.r_in))
        except: pass
        try: r += '\n{0:<15s} {1}'.format('r_out[1,m,1/m]', str(self.r_out))
        except: pass

        return r

class drift(marker):
    def __init__(self, pass_method = passmethods.pm_drift_pass, **kwargs):
        marker.__init__(self, pass_method = pass_method, **kwargs)

class corrector(marker):
    def __init__(self, pass_method = passmethods.pm_corrector_pass, hkick = None, vkick = None, **kwargs):
        marker.__init__(self, pass_method = pass_method, hkick = hkick, vkick = vkick, **kwargs)

class rfcavity(marker):
    def __init__(self, voltage = None, frequency = None, pass_method = passmethods.pm_cavity_pass, **kwargs):
        marker.__init__(self, pass_method = pass_method, voltage = voltage, frequency = frequency, energy = energy, hnumber = hnumber, **kwargs)

class quadrupole(marker):
    def __init__(self, pass_method = passmethods.pm_str_mpole_symplectic4_pass, nr_steps = 10, polynom_a = None, polynom_b = None, **kwargs):
        if 'thin_KL' in kwargs:
            if ('k' in kwargs) or (pass_method != passmethods.pm_thinquad_pass):
                raise Exception('Inconsistent marker parameters!')
        else:
            if polynom_a == None: polynom_a = [0.,0.,0.]
            if polynom_b == None: polynom_b = [0.,0.,0.]
            if 'k' in kwargs:
                polynom_b[1] = kwargs['k']
                del kwargs['k']
        marker.__init__(self, pass_method = pass_method, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)

class sextupole(marker):
    def __init__(self, pass_method = passmethods.pm_str_mpole_symplectic4_pass, nr_steps = 5, polynom_a = None, polynom_b = None, **kwargs):
        if polynom_a == None: polynom_a = [0,0,0]
        if polynom_b == None: polynom_b = [0,0,0]
        if 's' in kwargs:
            if ('sl' in kwargs) or (pass_method == 'thinsext_pass'):
                raise Exception('Inconsistent marker parameters!')
            polynom_b[2] = kwargs['s']
            del kwargs['s']
        if 'sl' in kwargs:
            if ('s' in kwargs) or (pass_method != 'thinsext_pass'):
                raise Exception('Inconsistent marker parameters!')
        marker.__init__(self, pass_method = pass_method, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)

class rbend(marker):
    def __init__(self, length, angle, pass_method = passmethods.pm_bnd_mpole_symplectic4_pass, nr_steps = 10, polynom_a = None, polynom_b = None, gap = 0, **kwargs):
        if polynom_a == None: polynom_a = [0,0,0]
        if polynom_b == None: polynom_b = [0,0,0]
        if 'k' in kwargs:
            polynom_b[1] = kwargs['k']
            del kwargs['k']
        if 's' in kwargs:
            polynom_b[2] = kwargs['s']
            del kwargs['s']
        if 'angle_in' not in kwargs:
            kwargs['angle_in'] = 0.5 * angle
        if 'angle_out' not in kwargs:
            kwargs['angle_out'] = 0.5 * angle
        marker.__init__(self, pass_method = pass_method, length = length, angle = angle, gap = gap, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)
